File: SalesTax/Golang/saleline.go
package main

import (
	"fmt"
	"math"
	"strings"
)

type SaleLine struct {
	Quantity       int
	ProductName    string
	Price          float64
	IsImported     bool
	LineValue, Tax float64
}

func (s SaleLine) ToString() string {
	return fmt.Sprintf("%d %s: %.2f", s.Quantity, s.ProductName, s.LineValue)
}

func NewSaleLine(quantity int, name string, price float64, isImported bool) SaleLine {
	var saleline SaleLine
	saleline.Quantity = quantity
	saleline.ProductName = name
	saleline.Price = price
	saleline.IsImported = isImported
	saleline.LineValue = price * float64(quantity)

	taxRate := 10

	if strings.Contains(name, "book") ||
		strings.Contains(name, "tablet") ||
		strings.Contains(name, "chip") {
		taxRate = 0
	}

	if isImported {
		taxRate = 5
	}

	taxAmt := CalculateTax(saleline.LineValue, taxRate)
	saleline.LineValue += taxAmt
	saleline.Tax = taxAmt

	return saleline
}

func CalculateTax(value float64, taxRate int) float64 {

	amount := roundPrecision((value*float64(taxRate))/100, 2)

	remainder := math.Remainder(amount, .05)

	if remainder > 0 {
		amount += .05 - remainder
	}
	return amount
}

func round(f float64) float64 {
	return math.Floor(f + .5)
}

func roundPrecision(f float64, places int) float64 {
	shift := math.Pow(10, float64(places))
	return round(f*shift) / shift
}

File: SalesTax/Typescript/saleline.ts
import * as numeral from 'numeral';

export default class SaleLine {
    private _productName: string;
    private _price: number;
    private _isImported: boolean;
    private _quantity: number;
    private _taxAmount: number;
    private _lineValue: number;

    get productName() {
        return this._productName;
    }

    get price() {
        return this._price;
    }

    get isImported() {
        return this._isImported;
    }

    get quantity() {
        return this._quantity;
    }

    get tax() {
        return this._taxAmount;
    }

    get lineValue() {
        return this._lineValue;
    }

    constructor(lineQuantity: number, name: string, unitPrice: number, itemIsImported: boolean) {
        let taxRate: number;
        
        if (!name) {
            name = '';
        }

        this._quantity = lineQuantity;
        this._productName = name;
        this._price = unitPrice;
        this._isImported = itemIsImported;
        this._lineValue = this.price * this.quantity;

        // calculate taxable amount
        // ideally should really have a product list and tax rules, but this'll have to do for the exercise.
        if (this.productName.indexOf('book') > -1 || this.productName.indexOf('tablet') > -1 || this.productName.indexOf('chip') > -1) {
            taxRate = 0 //No base tax applicable for books, medicals items or food
        } else {
            taxRate = 10 //10% base tax or general products
        } if (this.isImported) {
            taxRate = 5 //5% regardless for any imported items
        }

        this._taxAmount = SaleLine.calculateTax(this.lineValue, taxRate);
        //Add tax to line value
        this._lineValue += this._taxAmount;
        return;
    }

    static calculateTax(value: number, taxRate: number): number {
        let amount: number;
        let remainder: number;

        amount = Math.round((value * taxRate * 100)) / 100;

        //Now round up to nearest 5 cents.
        remainder = amount % .5;
        if (remainder > 0) {
            amount += .05 - remainder;
        }
        return amount;
    }

    toString(): string {
        return `${this.quantity} ${this.productName}: ${numeral(this.lineValue).format('0,0.00')}`;
    }
}